ckan/views/user.py

```diff
@@ -150,6 +151,92 @@ def read(id):
     return base.render(u'user/read.html', extra_vars)
 
 
+class ApiTokenView(MethodView):
+    def get(self, id, data=None, errors=None, error_summary=None):
+        context = {
+            u'model': model,
+            u'session': model.Session,
+            u'user': g.user,
+            u'auth_user_obj': g.userobj,
+            u'for_view': True,
+            u'include_plugin_extras': True
+        }
+        try:
+            tokens = logic.get_action(u'api_token_list')(
+                context, {u'user': id}
+            )
+        except logic.NotAuthorized:
+            base.abort(403, _(u'Unauthorized to view API tokens.'))
+
+        data_dict = {
+            u'id': id,
+            u'user_obj': g.userobj,
+            u'include_datasets': True,
+            u'include_num_followers': True
+        }
+
+        extra_vars = _extra_template_variables(context, data_dict)
+        if extra_vars is None:
+            return h.redirect_to(u'user.login')
+        extra_vars[u'tokens'] = tokens
+        extra_vars.update({
+            u'data': data,
+            u'errors': errors,
+            u'error_summary': error_summary
+        })
+        return base.render(u'user/api_tokens.html', extra_vars)
+
+    def post(self, id):
+        context = {u'model': model}
+
+        data_dict = logic.clean_dict(
+            dictization_functions.unflatten(
+                logic.tuplize_dict(logic.parse_params(request.form))))
+
+        data_dict[u'user'] = id
+        try:
+            token = logic.get_action(u'api_token_create')(
+                context,
+                data_dict
+            )[u'token']
+        except logic.NotAuthorized:
+            base.abort(403, _(u'Unauthorized to create API tokens.'))
+        except logic.ValidationError as e:
+            errors = e.error_dict
+            error_summary = e.error_summary
+            return self.get(id, data_dict, errors, error_summary)
+
+        copy_btn = dom_tags.button(dom_tags.i(u'', {
+            u'class': u'fa fa-copy'
+        }), {
+            u'type': u'button',
+            u'class': u'btn btn-default btn-xs',
+            u'data-module': u'copy-into-buffer',
+            u'data-module-copy-value': ensure_str(token)
+        })
+        h.flash_success(
+            _(
+                u'API Token created: <code style="word-break:break-all;">'
+                u'{token}</code> {copy}<br>'
+                u'Make sure to copy it now. '
+                u'It can be impossible to see it again!'
+            ).format(token=ensure_str(token), copy=copy_btn),
+            True
+        )
+        return h.redirect_to(u'user.api_tokens', id=id)
+
+
+def api_token_revoke(id, jti):
+    context = {u'model': model}
+    try:
+        logic.check_access(u'api_token_revoke', context, {u'token': jti})
+    except logic.NotAuthorized:
+        base.abort(403, _(u'Unauthorized to revoke API tokens.'))
+    model.ApiToken.revoke(jti)
```

Comment:
Should we wrap this into a `api_token_revoke` action? Even if it's a one line action. This will allow extensions to override it or chain it, otherwise it breaks our pattern of views  > action/auth > models

Reply:
Yes, updated